server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -144,111 +145,104 @@ public RepositoriesService(
+            .<RegisterRepositoryTaskResult>andThen((clusterUpdateStep, ignored) -> {
+                final ListenableFuture<AcknowledgedResponse> acknowledgementStep = new ListenableFuture<>();
+                final ListenableFuture<Boolean> publicationStep = new ListenableFuture<>(); // Boolean==changed.
+                submitUnbatchedTask(
+                    "put_repository [" + request.name() + "]",
+                    new RegisterRepositoryTask(this, request, acknowledgementStep) {
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.warn(() -> "failed to create repository [" + request.name() + "]", e);
+                            publicationStep.onFailure(e);
+                            super.onFailure(e);
+                        }
 
-            @Override
-            public boolean mustAck(DiscoveryNode discoveryNode) {
-                // repository is created on both master and data nodes
-                return discoveryNode.isMasterNode() || discoveryNode.canContainData();
-            }
+                        @Override
+                        public boolean mustAck(DiscoveryNode discoveryNode) {
+                            // repository is created on both master and data nodes
+                            return discoveryNode.isMasterNode() || discoveryNode.canContainData();
+                        }
 
-            @Override
-            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                if (changed) {
-                    if (found) {
-                        logger.info("updated repository [{}]", request.name());
-                    } else {
-                        logger.info("put repository [{}]", request.name());
+                        @Override
+                        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                            if (changed) {
+                                if (found) {
+                                    logger.info("updated repository [{}]", request.name());
+                                } else {
+                                    logger.info("put repository [{}]", request.name());
+                                }
+                            }
+                            publicationStep.onResponse(oldState != newState);
+                        }
                     }
+                );
+                publicationStep.addListener(clusterUpdateStep.delegateFailureAndWrap((ignored1, changed) -> {
+                    acknowledgementStep.addListener(clusterUpdateStep.delegateFailureAndWrap((ignored2, ack) -> {
+                        clusterUpdateStep.onResponse(new RegisterRepositoryTaskResult(ack, changed));
+                    }));
+                }));
+            })
+            .<AcknowledgedResponse>andThen((verificationStep, taskResult) -> {
+                if (request.verify() == false) {
```

Comment:
Think we can pull these conditions out here too?

```suggestion
                if (request.verify() == false || taskResult.ackResponse.isAcknowledged() == false || taskResult.changed == false) {
```

Reply:
I dont think it will have same meaning as before. In original code we do uuid update in "verify" block. I might have confusing names `verificationStep` and `verifyRepositoryStep` which might look similar

Original logic roughly this:
```
if verify:
  if ack&changed:
    verifyRepo
  updateUuid
  done
else
  done
```

Reply:
Oh yes you're quite right, thanks. Let's leave it as it is.